pkg/hubble/container/ring_reader.go

```diff
@@ -41,25 +44,43 @@ func NewRingReader(ring *Ring, start uint64) *RingReader {
 // Previous reads the event at the current position and decrement the read
 // position. When no more event can be read, Previous returns nil.
 func (r *RingReader) Previous() *v1.Event {
-	var e *v1.Event
 	// when the ring is not full, ring.read() may return <nil>, true
-	// in such a case, one should continue reading
-	for ok := true; e == nil && ok; r.idx-- {
-		e, ok = r.ring.read(r.idx)
+	// in such a case, there is no point in reading any further, since we
+	// have found the oldest entry.
+	e, ok := r.ring.read(r.idx)
+	if !ok || e == nil {
+		return nil
 	}
+	r.idx--
 	return e
 }
 
 // Next reads the event at the current position and increment the read position.
-// When no more event can be read, Next returns nil.
-func (r *RingReader) Next() *v1.Event {
-	var e *v1.Event
-	// when the ring is not full, ring.read() may return <nil>, true
-	// in such a case, one should continue reading
-	for ok := true; e == nil && ok; r.idx++ {
-		e, ok = r.ring.read(r.idx)
+// When no more event can be read, Next returns nil. If there was a concurrent
+// write, then we return an error.
+func (r *RingReader) Next() (*v1.Event, error) {
+	// When ok=false, there can be two reasons:
+	//
+	// Case 1 (reader ahead of writer): We have read past the writer. In this
+	// case, we want to return nil and don't bump the index, as we have read all
+	// existing values that exist now.
+	// Case 2 (writer ahead of reader): The writer has already overwritten the
+	// values we wanted to read. In this case, we return an error, as trying
+	// to catch up would be very racy.
+	//
+	// When the ring is not full, ring.read() may return ok=true but e=<nil>.
+	// Can treat this the same as we would treat ok=false, as
+	// `isReadAheadOfWriter` will still allow us to decide between which of the
+	// above cases occurred.
+	e, ok := r.ring.read(r.idx)
+	if !ok || e == nil {
+		if !r.ring.isReadAheadOfWriter(r.idx) {
+			return nil, ErrConcurrentWrite
+		}
+		return nil, nil // EOF
```

Comment:
`nil, nil` is a tricky return to handle properly on the consumer side. can we return `io.EOF` here instead?